doc/sfp-cmis/cmis-init.md

```diff
@@ -0,0 +1,301 @@
+
+## Warm Boot Requirements
+
+Functionality should continue to work across warm boot. 
+- The CMIS application initialization should not be performed during WarmBoot.
+- The CMIS application initialization should be skipped if no application code updates.
+
+# Functional Description
+
+- The **pmon#xcvrd** will detect the module type of the attched transceivers and post the
+information to the **STATE_DB**
+- If the transceiver is a CMIS module, the **pmon#xcvrd** will activate the appropriate
+application mode as per the current port mode.
+- If the transceiver is a CMIS module, the **show interfaces transceiver eeprom**
+should display the advertised applications of the transceiver.
+- If the transceiver is a CMIS module, the **sfputil** should be capable of detecting the errors of the CMIS application initialization.
+
+# Design
+
+## sonic-platform-daemons/sonic-xcvrd (modified)
+
+The transceiver daemon will be enhanced as below:
+
+- Post the CMIS application advertisement into the **STATE_DB**  
+  In the case of CMIS transceiver, the **application_advertisement** of **TRANSCEIVER_INFO**
+is updated from plain test into json dictionary format:  
+
+**Original:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "400GAUI-8 C2M (Annex 120E) - 400GBASE-DR4 (Cl 124)\n        100GAUI-2 C2M (Annex 135G) - 100G-FR/100GBASE-FR1 (Cl 140)",
+      ...... omitted ......
+    }
+  },
+```
+
+**Modified:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "{1: {'ap_sel_code_id': 1, 'host_electrical_interface_id': '400GAUI-8 C2M (Annex 120E)', 'module_media_interface_id': '400GBASE-DR4 (Cl 124)', 'host_lane_count': 8, 'media_lane_count': 4, 'host_lane_assignment_options': 1, 'media_lane_assignment_options': None}, 2: {'ap_sel_code_id': 2, 'host_electrical_interface_id': '100GAUI-2 C2M (Annex 135G)', 'module_media_interface_id': '100G-FR/100GBASE-FR1 (Cl 140)', 'host_lane_count': 2, 'media_lane_count': 1, 'host_lane_assignment_options': 85, 'media_lane_assignment_options': None}}",
+      ...... omitted ......
+    }
+  },
+  ...... omitted ......
+```  
+
+- Add **CMIS_MEDIA_SETTINGS** to the media_settings.json for custom CMIS controls
+
+**Format:**
+```
+{
+    "CMIS_MEDIA_SETTINGS":{
+        "<VENDOR_NAME>#<PART_NUMBER>": {
+            "application_selection": true|false
+        }
+    }
+}
+```
+
+**Example:**
+```
+{
+    "CMIS_MEDIA_SETTINGS":{
+        "AVAGO#AFCT-93DRPHZ-AZ2": {
+            "application_selection": true
+        },
+        "Amphenol#NDYYYF-0001": {
+            "application_selection": false
+        },
+        "DELL EMC#6MGDY": {
+            "application_selection": true
+        },
+        "DELL EMC#DH11M": {
+            "application_selection": false
+        }
+    }
+}
+```
+  As of now, only **application_selection** is supported, and it defaults to **True** if unspecified
```

Comment:
How does this interact with the module's advertising? I assume app selection would not be done for QSFP/SFP+ modules, but passive cables do not require app selection either.

Reply:
There are a couple of checkers in the CMIS Manager
1. Skip if the cage/form-factor type is not QSFPDD  (Or OSFP, but I'm not sure if it shares the same form-factor)
2. Skip if the media type (i.e. byte 128 or 0 of the module) is not a QSFPDD
3. Skip if the memory type of the module is FLAT
Upon port change events, as we support only dynamic-port-breakout, which means sub-ports will share the same config, hence the application code could be determined by the config of anyone of the sub-port in the PORT_SET events.

Reply:
And we also have another sanity check to skip the application initialization if
1. Defaults to application 1 if none of the applications matched the current speed and lane count of the host interface 
2. Skip the application init if no application code update while the current datapath is activated and module state is ready.